lib/puppet/http/client.rb

```diff
@@ -171,10 +171,40 @@ def execute_streaming(request, ssl_context:, user: nil, password: nil, &block)
     end
   end
 
+  def expand_into_parameters(data)
+    data.inject([]) do |params, key_value|
+      key, value = key_value
+
+      expanded_value = case value
+                       when Array
+                         value.collect { |val| [key, val] }
+                       else
+                         [key_value]
+                       end
+
+      params.concat(expand_primitive_types_into_parameters(expanded_value))
+    end
+  end
+
+  def expand_primitive_types_into_parameters(data)
+    data.inject([]) do |params, key_value|
+      key, value = key_value
+      case value
+      when nil
+        params
+      when true, false, String, Symbol, Integer, Float
+        params << [key, value]
+      else
+        raise ArgumentError, _("HTTP REST queries cannot handle values of type '%{klass}'") % { klass: value.class }
```

Comment:
@joshcooper is this what you were thinking? I just copied this code over. I still need to add test, and I feel like this can be cleaned up a bit too

Reply:
Maybe change the exception class to `Puppet::HTTP::Serialization`? Also it'd be good to add a test for this case, such as if a caller passes in an `Environment` object instead of a string.

Reply:
done